### PR TITLE
Introduce the last SRIOV patch-set

### PR DESCRIPTION
**Pull request by liuyuan-GitHub**
This patch-set is the last one for SRIOV feature, the previous patch-set
has implemented a SRIOV-capable device(PF) detection and interception,
a SRIOV VF creation. This patch-set mainly handles passthrough VF, disable VF.

v2) 1. introduce member zombie for pci_vdev structure
    2. rebase the patch-set

v3) 1. refine VF BARs and MSIx initialization
    2. refine comments and commit message
    3. rebase the code

v4) 1. move VF BARs initialization from init_sriov_vf_vdev to create_vf
    2. rebase code for assigned VF emulation
    3. refine coding style and comments

v5) 1. move assigned VF cfg space emulation to init_vdev_pt
    2. refine code style

Tracked-On: #4433

Signed-off-by: Yuan Liu <yuan1.liu@intel.com>
Acked-by: Eddie Dong <eddie.dong@intel.com>

### CODE REVIEW

**Comment by liuyuan-GitHub** [comment]
@dongyaozu  @lifeix 

I add a new patch "hv: vdev uses vpci member to check which VM it belongs to" to fix a bug, please help to review in here

If there is a vdev which is already zombie instance(vdev->vpci is NULL), the HV will crash when a post-launched VM shutdown.

**Comment by HuiHuangShi** [comment]
37 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:sharing_config filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_unmap_msix offset:40<br> reason:Pointer param should be declared pointer to const. : vdev<br><2> config:part_config filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_unmap_msix offset:40<br> reason:Pointer param should be declared pointer to const. : vdev<br><3> config:part_config filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_map_msix offset:69<br> reason:Pointer param should be declared pointer to const. : vdev<br><4> config:sharing_config filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_map_msix offset:69<br> reason:Pointer param should be declared pointer to const. : vdev<br><5> config:common filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_map_msix offset:84<br> reason:Casting operation to a pointer. : (void* to unsigned long*): ( uint64_t * ) vm -> arch_vm.nworld_eptp<br><6> config:common filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_map_msix offset:84<br> reason:Casting operation on a pointer. : (void* to unsigned long*): ( uint64_t * ) vm -> arch_vm.nworld_eptp<br><7> config:common filename:/hypervisor/dm/vpci/pci_pt.c func:init_vdev_pt offset:376<br> reason:For loop incrementation is not simple.<br><8> config:sharing_config filename:/hypervisor/dm/vpci/pci_pt.c func:init_vdev_pt offset:382<br> reason:Potential side effect problem in expression. : (vdev)->nr_bars<br><9> config:part_config filename:/hypervisor/dm/vpci/pci_pt.c func:init_vdev_pt offset:382<br> reason:Potential side effect problem in expression. : (vdev)->nr_bars<br><10> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:create_vf offset:123<br> reason:Pointer not checked for null before use. : vf_vbar<br><11> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:create_vf offset:123<br> reason:Pointer not checked for null before use. : vf_vbar<br><12> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:create_vf offset:124<br> reason:Pointer not checked for null before use. : vf_vdev<br><13> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:create_vf offset:124<br> reason:Pointer not checked for null before use. : vf_vdev<br><14> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:186<br> reason:Function return value potentially unused. : num_vfs<br><15> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:186<br> reason:Function return value potentially unused. : num_vfs<br><16> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:196<br> reason:Function return value potentially unused. : vf_bdf.fields.bus<br><17> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:196<br> reason:Function return value potentially unused. : vf_bdf.fields.bus<br><18> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:197<br> reason:Function return value potentially unused. : vf_bdf.fields.devfun<br><19> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:197<br> reason:Function return value potentially unused. : vf_bdf.fields.devfun<br><20> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:228<br> reason:Pointer param should be declared pointer to const. : pf_vdev<br><21> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:228<br> reason:Pointer param should be declared pointer to const. : pf_vdev<br><22> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:245<br> reason:Function return value potentially unused. : first<br><23> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:245<br> reason:Function return value potentially unused. : first<br><24> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:246<br> reason:Function return value potentially unused. : stride<br><25> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:246<br> reason:Function return value potentially unused. : stride<br><26> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:250<br> reason:Function return value potentially unused. : bdf.fields.bus<br><27> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:250<br> reason:Function return value potentially unused. : bdf.fields.bus<br><28> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:251<br> reason:Function return value potentially unused. : bdf.fields.devfun<br><29> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:251<br> reason:Function return value potentially unused. : bdf.fields.devfun<br><30> config:part_config filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:351<br> reason:Pointer param should be declared pointer to const. : vpci<br><31> config:sharing_config filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:351<br> reason:Pointer param should be declared pointer to const. : vpci<br><32> config:part_config filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:360<br> reason:Local structure returned in function result. : vdev<br><33> config:sharing_config filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:360<br> reason:Local structure returned in function result. : vdev<br><34> config:common filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:360<br> reason:Pointer assignment to wider scope.<br>.....<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

File: hypervisor/dm/vpci/pci_pt.c [hunk 1]
```diff
@@ -387,5 +387,17 @@ void init_vdev_pt(struct pci_vdev *vdev, bool is_pf_vdev)
 			pci_command |= 0x400U;
 			pci_pdev_write_cfg(vdev->pdev->bdf, PCIR_COMMAND, 2U, pci_command);
 		}
+	} else {
+		if (vdev->vpci != vdev->phyfun->vpci) {
```
Comment (dongyaozu) [inline hypervisor/dm/vpci/pci_pt.c]: /* VF in a UOS */

File: hypervisor/dm/vpci/vpci.c [hunk 2]
```diff
@@ -721,10 +727,18 @@ int32_t vpci_assign_pcidev(struct acrn_vm *tgt_vm, struct acrn_assign_pcidev *pc
 			vdev_in_sos->vpci = vpci;
 
 			spinlock_obtain(&tgt_vm->vpci.lock);
-			vdev = vpci_init_vdev(vpci, vdev_in_sos->pci_dev_config, NULL);
+			vdev = vpci_init_vdev(vpci, vdev_in_sos->pci_dev_config, vdev_in_sos->phyfun);
 			pci_vdev_write_vcfg(vdev, PCIR_INTERRUPT_LINE, 1U, pcidev->intr_line);
 			pci_vdev_write_vcfg(vdev, PCIR_INTERRUPT_PIN, 1U, pcidev->intr_pin);
 			for (idx = 0U; idx < vdev->nr_bars; idx++) {
+				/* VF physical BARs are zero & RO, initialze assigned VF BARs from SOS */
```
Comment (dongyaozu) [inline hypervisor/dm/vpci/vpci.c]: This comment seems to be irrelevant, is it? Delete this line of comment?

File: hypervisor/dm/vpci/vpci.c [hunk 3]
```diff
@@ -721,10 +727,18 @@ int32_t vpci_assign_pcidev(struct acrn_vm *tgt_vm, struct acrn_assign_pcidev *pc
...
+				if (vdev->phyfun != NULL) {
```
Comment (dongyaozu) [inline hypervisor/dm/vpci/vpci.c]: /* VF is assigned to a UOS */

File: hypervisor/dm/vpci/pci_pt.c (hunk 1 above)
Comment (liuyuan-GitHub) [reply hypervisor/dm/vpci/pci_pt.c]: Sure, I add comment in here

File: hypervisor/dm/vpci/vpci.c (hunk 2 above)
Comment (liuyuan-GitHub) [reply hypervisor/dm/vpci/vpci.c]: OK, I will remove it

File: hypervisor/dm/vpci/vpci.c (hunk 3 above)
Comment (liuyuan-GitHub) [reply hypervisor/dm/vpci/vpci.c]: Got it, I will update the comment

**Comment by dongyaozu** [comment]
@liuyuan-GitHub can we merge the 9/9 patch with the one you bring disable_vfs where the zombie state is introduced ?

**Comment by HuiHuangShi** [comment]
36 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:sharing_config filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_unmap_msix offset:40<br> reason:Pointer param should be declared pointer to const. : vdev<br><2> config:part_config filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_unmap_msix offset:40<br> reason:Pointer param should be declared pointer to const. : vdev<br><3> config:part_config filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_map_msix offset:69<br> reason:Pointer param should be declared pointer to const. : vdev<br><4> config:sharing_config filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_map_msix offset:69<br> reason:Pointer param should be declared pointer to const. : vdev<br><5> config:common filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_map_msix offset:84<br> reason:Casting operation to a pointer. : (void* to unsigned long*): ( uint64_t * ) vm -> arch_vm.nworld_eptp<br><6> config:common filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_map_msix offset:84<br> reason:Casting operation on a pointer. : (void* to unsigned long*): ( uint64_t * ) vm -> arch_vm.nworld_eptp<br><7> config:common filename:/hypervisor/dm/vpci/pci_pt.c func:init_vdev_pt offset:376<br> reason:For loop incrementation is not simple.<br><8> config:sharing_config filename:/hypervisor/dm/vpci/pci_pt.c func:init_vdev_pt offset:382<br> reason:Potential side effect problem in expression. : (vdev)->nr_bars<br><9> config:part_config filename:/hypervisor/dm/vpci/pci_pt.c func:init_vdev_pt offset:382<br> reason:Potential side effect problem in expression. : (vdev)->nr_bars<br><10> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:create_vf offset:123<br> reason:Pointer not checked for null before use. : vf_vbar<br><11> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:create_vf offset:123<br> reason:Pointer not checked for null before use. : vf_vbar<br><12> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:create_vf offset:124<br> reason:Pointer not checked for null before use. : vf_vdev<br><13> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:create_vf offset:124<br> reason:Pointer not checked for null before use. : vf_vdev<br><14> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:196<br> reason:Function return value potentially unused. : vf_bdf.fields.bus<br><15> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:196<br> reason:Function return value potentially unused. : vf_bdf.fields.bus<br><16> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:197<br> reason:Function return value potentially unused. : vf_bdf.fields.devfun<br><17> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:197<br> reason:Function return value potentially unused. : vf_bdf.fields.devfun<br><18> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:229<br> reason:Pointer param should be declared pointer to const. : pf_vdev<br><19> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:229<br> reason:Pointer param should be declared pointer to const. : pf_vdev<br><20> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:246<br> reason:Function return value potentially unused. : first<br><21> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:246<br> reason:Function return value potentially unused. : first<br><22> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:247<br> reason:Function return value potentially unused. : stride<br><23> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:247<br> reason:Function return value potentially unused. : stride<br><24> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:251<br> reason:Function return value potentially unused. : bdf.fields.bus<br><25> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:251<br> reason:Function return value potentially unused. : bdf.fields.bus<br><26> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:252<br> reason:Function return value potentially unused. : bdf.fields.devfun<br><27> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:252<br> reason:Function return value potentially unused. : bdf.fields.devfun<br><28> config:part_config filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:351<br> reason:Pointer param should be declared pointer to const. : vpci<br><29> config:sharing_config filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:351<br> reason:Pointer param should be declared pointer to const. : vpci<br><30> config:part_config filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:360<br> reason:Local structure returned in function result. : vdev<br><31> config:sharing_config filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:360<br> reason:Local structure returned in function result. : vdev<br><32> config:common filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:360<br> reason:Pointer assignment to wider scope.<br><33> config:common filename:/hypervisor/dm/vpci/vpci.c func:deinit_prelaunched_vm_vpci offset:646<br> reason:Space missing before or after binary operator.<br><34> config:common filename:/hypervisor/dm/vpci/vpci.c func:deinit_postlaunched_vm_vpci offset:686<br> reason:Space missing before or after binary operator.<br>.....<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>

**Comment by ClaudZhang1995** [comment]
OK to verify

**Comment by HuiHuangShi** [comment]
36 new potential violations to the coding guideline detected. Please help check. Thanks!<br><1> config:sharing_config filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_unmap_msix offset:40<br> reason:Pointer param should be declared pointer to const. : vdev<br><2> config:part_config filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_unmap_msix offset:40<br> reason:Pointer param should be declared pointer to const. : vdev<br><3> config:part_config filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_map_msix offset:69<br> reason:Pointer param should be declared pointer to const. : vdev<br><4> config:sharing_config filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_map_msix offset:69<br> reason:Pointer param should be declared pointer to const. : vdev<br><5> config:common filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_map_msix offset:84<br> reason:Casting operation to a pointer. : (void* to unsigned long*): ( uint64_t * ) vm -> arch_vm.nworld_eptp<br><6> config:common filename:/hypervisor/dm/vpci/pci_pt.c func:vdev_pt_map_msix offset:84<br> reason:Casting operation on a pointer. : (void* to unsigned long*): ( uint64_t * ) vm -> arch_vm.nworld_eptp<br><7> config:common filename:/hypervisor/dm/vpci/pci_pt.c func:init_vdev_pt offset:376<br> reason:For loop incrementation is not simple.<br><8> config:sharing_config filename:/hypervisor/dm/vpci/pci_pt.c func:init_vdev_pt offset:382<br> reason:Potential side effect problem in expression. : (vdev)->nr_bars<br><9> config:part_config filename:/hypervisor/dm/vpci/pci_pt.c func:init_vdev_pt offset:382<br> reason:Potential side effect problem in expression. : (vdev)->nr_bars<br><10> config:common filename:/hypervisor/include/arch/x86/guest/vm.h func:is_own_device offset:217<br> reason:Static procedure is not explicitly called in code analysed. : is_own_device<br><11> config:common filename:/hypervisor/include/arch/x86/guest/vm.h func:is_zombie_vf offset:229<br> reason:Static procedure is not explicitly called in code analysed. : is_zombie_vf<br><12> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:create_vf offset:123<br> reason:Pointer not checked for null before use. : vf_vbar<br><13> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:create_vf offset:123<br> reason:Pointer not checked for null before use. : vf_vbar<br><14> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:create_vf offset:124<br> reason:Pointer not checked for null before use. : vf_vdev<br><15> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:create_vf offset:124<br> reason:Pointer not checked for null before use. : vf_vdev<br><16> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:196<br> reason:Function return value potentially unused. : vf_bdf.fields.bus<br><17> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:196<br> reason:Function return value potentially unused. : vf_bdf.fields.bus<br><18> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:197<br> reason:Function return value potentially unused. : vf_bdf.fields.devfun<br><19> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:enable_vfs offset:197<br> reason:Function return value potentially unused. : vf_bdf.fields.devfun<br><20> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:229<br> reason:Pointer param should be declared pointer to const. : pf_vdev<br><21> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:229<br> reason:Pointer param should be declared pointer to const. : pf_vdev<br><22> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:246<br> reason:Function return value potentially unused. : first<br><23> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:246<br> reason:Function return value potentially unused. : first<br><24> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:247<br> reason:Function return value potentially unused. : stride<br><25> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:247<br> reason:Function return value potentially unused. : stride<br><26> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:251<br> reason:Function return value potentially unused. : bdf.fields.bus<br><27> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:251<br> reason:Function return value potentially unused. : bdf.fields.bus<br><28> config:sharing_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:252<br> reason:Function return value potentially unused. : bdf.fields.devfun<br><29> config:part_config filename:/hypervisor/dm/vpci/vsriov.c func:disable_vfs offset:252<br> reason:Function return value potentially unused. : bdf.fields.devfun<br><30> config:part_config filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:351<br> reason:Pointer param should be declared pointer to const. : vpci<br><31> config:sharing_config filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:351<br> reason:Pointer param should be declared pointer to const. : vpci<br><32> config:part_config filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:360<br> reason:Local structure returned in function result. : vdev<br><33> config:sharing_config filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:360<br> reason:Local structure returned in function result. : vdev<br><34> config:common filename:/hypervisor/dm/vpci/vpci.c func:find_available_vdev offset:360<br> reason:Pointer assignment to wider scope.<br>.....<br>No New Name Conflict<br><br>No New Function Declaration/Definition Mismatch<br>
